### ISSUE TITLE
Allow RegExp or wildcard for ingress domain and/or path

### ISSUE THREAD

**Issue by deitch** [kubernetes/kubernetes#41881]
**Is this a request for help?** 

No

**What keywords did you search in Kubernetes issues before filing this one?** (If you have found any duplicates, you should instead reply there.):

* ingress wildcard
* ingress regexp

**Is this a BUG REPORT or FEATURE REQUEST?**

Feature, although if it exists and I don't see it, then bug for docs

---

As far as I can tell, k8s ingress supports (in theory), subdomain wildcards for ingresses (supposing the given controller supports them, of course), such as `*.mydomain.com`. There appear to be confusing issues (re: https://github.com/kubernetes/kubernetes/issues/39622 ) but in theory it supports it.

1. I cannot find this in official documentation anywhere?
2. Is there support for general wildcards?

The specific use case is as follows. I have a set of k8s resources being deployed to multiple environments, e.g. prod, uat, qa. Each one has its own inbound root domain, e.g. `prod.mydomain.com`, `qa.mydomain.com`, `uat.mydomain.com`.

When I create my k8s resources, I want to use identical ones between all the environments. If I have 3 exposed services (e.g. `web`, `api`, `magic`) in each domain, routed by subdomain, I shouldn't require 3 separate ingresses, one for each environment, per exposed service. Instead, I want to do something like this:

```yml
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  name: web
spec:
  rules:
  - host: "web.*"
    http:
      paths:
      - path: /
        backend:
          serviceName: web-service
          servicePort: 80
```

(repeat for services `api` and `magic`)

This allows my to tell my devs

1. Pick your environment with `kubectl config set-context prod` (or `uat` or `qa`)
2.Deploy consistently with `kubectl apply -f configs/`

Did I miss that somewhere? Or is it not currently supported?

**Comment by cmluciano** [comment]
wildcard support is in kubernetes/ingress#8

**Comment by deitch** [comment]
@cmluciano thanks. I saw that earlier, but didn't think it answers, because I thought:

1. It mainly refers to the nginx implementation as an ingress controller, not the ingress spec.
2. It allows wildcards as _sub_domain, but doesn't appear to allow for the _parent_ domain.

I want to be able to do

```yml
host: "web.*"
```

And thus have any request whose host header starts with `web.` match.

**Comment by deitch** [comment]
@cmluciano and tried it, got an invalid spec error.

**Comment by hjacobs** [comment]
I think the Ingress spec should clearly state how to treat wildcards/regexp. We have [our own Ingress](https://kubernetes-on-aws.readthedocs.io/en/latest/admin-guide/kubernetes-in-production.html#ingress) controller and HTTP proxy ([Skipper](https://github.com/zalando/skipper)), i.e. we would like to have a clear spec to implement (and not do our own potentially incompatible extension).

**Comment by deitch** [comment]
@hjacobs I looked at the source code, and it seems only to allow it as the left-most element.

Personally, I would be content to have `host: myhost` and have the ingress controller know to append `.mysub.mydomain.com`. But I have yet to find a controller that supports that.

**Comment by klausenbusk** [comment]
+1 for this feature, although I don't need regex. I just want a way that I can specific multiple hosts for a ingress.

Like: https://github.com/kubernetes/ingress/issues/87
> Nginx supports it, http://nginx.org/en/docs/http/ngx_http_core_module.html#server_name would remove quite a bit of boilerplate in my app. I would make a PR but my competency is not Go.
> 
> apiVersion: extensions/v1beta1
> kind: Ingress
> metadata:
>   name: web-ingress
> spec:
>   rules:
>     hosts:
>       - 1.myawesomesite.co
>       - 2.myawesomesite.co
>     http:
>       paths:
>       - path: /
>         backend:
>           serviceName: awesomesite-web
>           servicePort: 80

**Comment by deitch** [comment]
No additional thoughts on this?

**Comment by rdsubhas** [comment]
Yep, this is somewhat essential for using minikube or kubernetes from dev-to-prod. We have dozens of developers using minikube and hosted/remote kubernetes VMs with IPs. Everyone cannot be given full domain names (/etc/hosts overrides and generally converting this into a huge problem).

Almost all routers that I know of (haproxy, nginx, etc) allow routing based on trailing wildcards, i.e. only the subdomain matters, and the domain doesn't. But here, its reverse, the domain seems to matter and subdomain is allowed a wildcard, which I don't know but seems to be a not-so-common routing policy. I believe this is because Ingress was mostly designed with some specific cloud limitations.

Would make a lot of sense to expose trailing wildcards, and allow the Ingress controller to take the final call whether its valid or not.

**Comment by deitch** [comment]
Thanks @rdsubhas ; well explained.

**Comment by rdsubhas** [comment]
@deitch until then if anyone needs to do thins in the official nginx ingress, here is a shortcut:

- Set `host: <just the subdomain>` in Ingress rule, e.g. `host: myapp`
- And then use an init container to replace `server_name {{$server.Hostname}};` in the default nginx template with `server_name {{$server.Hostname}} {{$server.Hostname}}.*;`

**Comment by deitch** [comment]
@rdsubhas an init container where? In the nginx Deployment / DaemonSet (depending on your deployment preference)? I kind of did something similar with traefik, changing their template. Still, it shouldn't be _this_ hard.

**Comment by robermorales** [comment]
I used this container in the past with this purpose, and it did the job.

https://hub.docker.com/r/jwilder/nginx-proxy/

**Comment by deitch** [comment]
Any updates here?

@rdsubhas I am using traefik in one use case (nginx in another), and I did something similar. Actually, since nginx uses go templates, I was able to make the transformation right in the template itself. 

The problem I now have is trying to use letsencrypt, most plugins (like traefik's built-in acme support or kube-lego) pull the `Host:` right out of the `Ingress` resource (probably correctly). That does mean that any such transformation is going to be missed, since it happens later. Now I need to build support in for that.

Is there any way to create "automatic transform on demand" for resources? E.g. my cluster will transform the `Ingress` resource upon receipt to modify the `Host:` section?

**Comment by debianmaster** [comment]
+1 for this

**Comment by kedare** [comment]
I confirm this is a huge issue for us.
What is sad is that our ingress controller does support regex matching or wildcard at any position of the domain, but the configuration gets blocked because it doesn't pass validation at Kubernetes level.
Maybe there would be a flag like rawHost or bypassHostValidation to directly pass the value to the ingress controller without any kind of restrictions from Kubernetes ?

**Comment by deitch** [comment]
Ours does as well. Well, we use traefik, and no big deal to change the config to support it.

As a workaround, we use a string that never would be used anywhere in out `.yml` files (`anydomain123` for what it's worth), and then have our traefik config parse and replace that with a wildcard.

> Maybe there would be a flag like rawHost or bypassHostValidation

Or, just let it go through no matter what.

**Comment by Preskton** [comment]
This is extremely important to us as well -- our user story goes something like this: 

> As an Application Engineer, I want ingress rules to match on the start of a host pattern (like `foo.*` so that I can use the same ingress definition across environments (`foo.dev`, `foo.qa`, etc.) without transformation so that the risk of error switching across environments is reduced significantly. Without this feature, I'm not deploying the same thing to every environment, which can lead to errors and confusion during cluster operations and deployments.

There's a talk about "the spec" and "the code" in this thread - I could probably figure out the code, but how does "the spec" get approved/changed/whatever'd so that we can update the code to match our desired behavior?

**Comment by deitch** [comment]
I, too, would be happy to do a PR, but only if someone can point me into what is needed for the spec to change and then the code, and more importantly that it is supported in principle.

**Comment by DavidWylie** [comment]
This is a feature which allows ingress to be useful.  Without it the yaml definition file for an app (ingress + deployment + service)  cannot be environment independant.  Without it you end up with a proxy application / loadbalancer infront of the ingress just to change the url to a common non routable value.  


just supporting the * in the domain as a single part would be sufficient.  ie
svc.*.domain

**Comment by rdsubhas** [comment]
I think, from a Spec perspective, having it as a FQDN helps a lot of "standard/kubernetes compliant" resources to happen. Although the API is called `Ingress`, this has a wider impact than just routing. Because if the domain is unspecified, then things like issuing letsencrypt certificates, or creating DNS names for services, etc. are going to become really hard. Different Kubernetes cloud providers would start interpreting the spec differently if its no longer a FQDN, which _could_ lead to fragmentation (wildcards vs regexes, PCRE regex vs POSIX regex, etc).

While its true that development is hard, supporting domain prefixes _could_ be done today. If we specify the `host: foo` - it is indeed considered a valid FQDN. We can instrument the ingress controller (or dns/ssl controllers) to automatically interpret the host as `foo.<real-domain>`. As a cluster admin, we presently patch the nginx ingress controller to add our production suffixes automatically if not present, and our engineers only use subdomains.

Just two cents, maybe pressuring downstream API controllers (like dns/ssl/ingress controllers, and maybe google cloud ingress config ;) to support suffixes/prefixes to the host is a better/longer term solution, than making the spec itself widely open to interpretation and being stringly typed.

**Comment by DavidWylie** [comment]
@rdsubhas   Looking into this in the issues form some of the implementors of the ingress controller spec.  There almost standard response is its not in the spec so we dont want to differ from it.  
Even just a standard field or annotation would be enough to get some of them moving in a shared direction.    you seem to have hit the nail on the head with the foo.<real-domain>  with that the real domain is a seperate piece of data which as an optional thing could be a string or loaded form a config map.

**Comment by deitch** [comment]
Looks like I stirred up a bit of a hornet's nest with this one... which means it matters to people.

Perhaps the strongest sign is that every time I interact with someone who deploys to kubernetes clusters with an automated CD pipeline across multiple environments, we *always* end up discussing, "so how did you hack around the Ingress `Host:` limitation?" If so many people are doing it, then it already is fragmenting, but at a per-deployment level. Fragmenting across ingress controllers might be an improvement. :-)

If one of the key goals of kubernetes is making services easy to deploy and - "automating devops" like Kelsey's interview last week, if you will - then the goal is practical: how do I make deploying services config-driven, reliable, reproducible, scalable. Requiring the config on each exposed service to be different from one environment to another makes an automated CD pipeline brittle and extremely difficult (and hacked and fragmented). I need to maintain separate config files, when they differ only by the ending of the domain.

I am unclear is to why the spec for an FQDN **must** be the spec for specifying what service `X` should respond to? (I also just violated the spec of not ending a sentence with a preposition, oops :-) )

Agreed wholeheartedly, it needs a standard, and the current one is forcing people to break it in implementation. I _do_ think that using a standard that wasn't written for specifying services deployed and identified dynamically in multiple environments is the best one to use. If there isn't a better, let's define one.

> maybe pressuring downstream API controllers

Isn't the way to pressure `IngressController` creators by defining a spec?

In any case, I much appreciate the feedback and time spent on this issue.
If you want to call it `HostRegex:`, that works, too. Call it `HostPrefix`, or anything you want.

**Comment by Preskton** [comment]
I like all this discussion: this is fun. :)

As a newbie here, it feels like "changing the spec" is a really big deal, as I'm sure there are all kinds of downstream implications, like @rdsubhas implies. Outside the current discussion, I have a couple of more meta-questions:

- Where is "the spec"? I assume there's something deeper than https://kubernetes.io/docs/concepts/services-networking/ingress/?
- How would someone introduce a proposed change to "the spec"? Is it strictly via the SIGs?
- Who are the "`IngressController` creators? How does a change to the "the spec" get voted on/ratified/whatever?
- When is it preferable to split out a new annotation vs. a new property?

This is pretty offtopic, so happy to take it to DMs/Slack if that makes more sense.

**Comment by andrewtar** [comment]
Are there any results or changes for this issue? We really need to use * to specify a set of subdomains. The approach with listing of all possible subdomains is not suitable for us, because the number of subdomains is too large. It was very nice to use *. We found an [alternative implementation of IngressController](https://docs.traefik.io/user-guide/kubernetes/), but we can not use this, because our entire infrastructure is already built on Nginx Ingress.

**Comment by JannikZed** [comment]
Same here, we would love to see the wildcard in the path variable like `/index-*` ..

**Comment by santiagopoli** [comment]
+ 1

**Comment by licarth** [comment]
@rdsubhas Here is my implementation of what you suggedted, with `nginx-ingress` helm chart :

`values.yaml`
```
controller:
  extraInitContainers:
    - name: custom-template
      image: "quay.io/kubernetes-ingress-controller/nginx-ingress-controller:0.15.0"
      command: ["/bin/bash","-c"]
      args: ["sed -e 's/server_name {{ \\$server.Hostname }}/server_name {{ $server.Hostname }} {{ $server.Hostname }}.*/g' /etc/nginx/template/nginx.tmpl > /templates/nginx.tmpl"]
      volumeMounts:
      - mountPath: /templates
        name: nginx-template-volume
  extraVolumeMounts: 
    - mountPath: /etc/nginx/template
      name: nginx-template-volume
      readOnly: true
  extraVolumes:
    - name: nginx-template-volume
      emptyDir: {}
```
then 
```
helm upgrade --install nginx-ingress stable/nginx-ingress -f values.yaml
```

**Comment by haizaar** [comment]
Can please someone un-stale it?

**Comment by WoLfulus** [comment]
+1 for this... I'm trying to map `*.*.domain` and I can't :(

**Comment by swoldemi** [comment]
Does https://kubernetes.github.io/ingress-nginx/user-guide/ingress-path-matching/ reference the desired behaviour that this PR requests? (For paths, not domains)

**Comment by deitch** [comment]
@swoldemi possibly, but:

1. It doesn't support domains, which is the larger issue to handle deployment to multiple environments.
2. it is specific to nginx, and thus not an official part of the `Ingress` api

**Comment by himao** [comment]
Any update?
I need `Host: test-*.mydomain.com`,but it doesn't work because it doesn't pass validation at Kubernetes level.
I'm use the traefik controller,or any other way?

**Comment by travisghansen** [comment]
@himao it's not a 'supported' way but with `traefik` ingress controller I simply enabled the `file` backend and then manually created the following snippet/file:

```
[backends]
  [backends.gitlab-pages]
    [backends.gitlab-pages.servers]
      [backends.gitlab-pages.servers.server0]
        url = "http://gitlab.default:80"
        weight = 1

[frontends]
  [frontends.gitlab-pages]
    backend = "gitlab-pages"
    passHostHeader = true
  [frontends.gitlab-pages.routes]
    [frontends.gitlab-pages.routes.route0]
      rule = 'HostRegexp:{subdomain:[A-Za-z0-9\-]+}.pages.foodomain.com'
```

While I'm using a persistent volume to mount the folder containing my `rules` files you could easily use a `ConfigMap` volume and manage everything with cluster resources.  I've been doing this for a couple years now and never had an issue with it.  It's not super 'clean' but it works for me.

**Comment by himao** [comment]
@travisghansen thank u,i will try it.

**Comment by deitch** [comment]
I like it @travisghansen . But as you said, "it's not super clean." This thing really should be.

**Comment by shinebayar-g** [comment]
@travisghansen Does sticky session works when using this method?

**Comment by travisghansen** [comment]
@deitch yeah, I unfortunately don't think it's going with the `Ingress` spec.  As some of the teams are working on `IngressRoute` as `CRD`s maybe some day the next-gen spec will support it.

**Comment by travisghansen** [comment]
@shinebayar-g not really.  The example has a single backend pointing to the kubernetes `Service` so as far a `traefik` is concerned it's a single endpoint.  If you have stable pod names (ie: `StatefulSet`) you could do sticky sessions but otherwise that's one of the (several) downsides.

**Comment by thockin** [comment]
Ingress is sort of lowest-common-API.  As such it will probably not get more than simple prefix matching.  The KEP to move to GA details some places we can clean up the spec, but it wil not get much more featureful.

https://github.com/kubernetes/enhancements/blob/master/keps/sig-network/20190125-ingress-api-group.md

There will be a followup APi which has possibly more capabilities.

**Comment by desmatz** [comment]
There is an annotation for ingress object: 
https://kubernetes.github.io/ingress-nginx/user-guide/nginx-configuration/annotations/#server-alias
```
nginx.ingress.kubernetes.io/server-alias: "<alias>"
```
So ingress object will be:
```yaml
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  name: {{ some name }}
  annotations:
    kubernetes.io/ingress.class: "nginx"
    nginx.ingress.kubernetes.io/force-ssl-redirect: "true"
    nginx.ingress.kubernetes.io/server-alias: "*.{{ domain }}"
spec:
  tls:
  - hosts: 
    - {{ domain }}
  rules:
  - host: {{ domain }}
    http:
      paths:
      - path: /
        backend:
          serviceName: {{ svc name }}
          servicePort: {{ svc port }}
```
And server name in the nginx.conf will be
```
server_name {{ domain }} *.{{ domain }};
```

**Comment by deitch** [comment]
@desmatz that works well for nginx, but it is a specific implementation, not API standard, and thus won't work with other `IngressController`s

**Comment by deitch** [comment]
@thockin wrote:

> There will be a followup API which has possibly more capabilities.

Is there information about it?

**Comment by adgang** [comment]
> Ingress is sort of lowest-common-API. As such it will probably not get more than simple prefix matching. The KEP to move to GA details some places we can clean up the spec, but it wil not get much more featureful.

I agree. But some regex in DNS is also lowest common API. It would be horrible to support stuff like 
abc-*.foo.com. But host names like *.foo.com and supporting domain prefixes seem to be reasonable features.

https://www.ietf.org/rfc/rfc4592.txt

**Comment by mrmuli** [comment]
I recently had to capture the subdomain while we can't do regex on spec. I ended up adding a server alias of the regex as below, then used a different domain I own as a rule such that nginx always matches the regex and avails the variable

```
annotations:
    nginx.ingress.kubernetes.io/server-alias: '~^(?<subdomain>.+)\.example\.com$'
```


```
rules:
  - host: "*.my-reserved-name.com"
```

So the nginx config will always match my regex, while my reserved name is skipped
```
server {
    server_name *.my-reserved-name.com ~^(?<subdomain>.+)\.example\.com$;
```

**Comment by demisx** [comment]
IMHO, this is should be reopened. We need support at least for listing multiple hosts or specifying glob-like wildcards, if regex is not feasible.

**Comment by rikatz** [comment]
@deitch Kubernetes 1.18 has been released with some enhancements, as you can see here: https://kubernetes.io/blog/2020/04/02/improvements-to-the-ingress-api-in-kubernetes-1.18/

One of them is the support for wildcard in hostname, and a better handling of Path.

As a remember that this an API change and also needs some attention of each ingress controller implementation, do you think it's enough and the issue can be closed?

Thank you

**Comment by deitch** [comment]
Hi @rikatz . Unless I misunderstood the release notes (thanks for referencing them here; I read the k8s 1.18 release notes in general, and missed the ingress part. Oops. :-) ), it doesn't do it.

The wildcard support listed there is `*.foo.com`, while this is looking for the exact opposite, `api.*`. However, if the wildcard support is more generalized and I misunderstood, such that `api.*` and `web.*` or even `web.*.foo.com` would work, then, yes, indeed, it solves it.

Do you mind clarifying?

**Comment by rikatz** [comment]
No, it does not work.

The idea of wildcard is to have a catchall for a specific domain like *.yoursite.com, so I don't think it covers the original case

**Comment by deitch** [comment]
Then I think we should leave this open. Here was from the original issue:

> The specific use case is as follows. I have a set of k8s resources being deployed to multiple environments, e.g. prod, uat, qa. Each one has its own inbound root domain, e.g. prod.mydomain.com, qa.mydomain.com, uat.mydomain.com.
>
> When I create my k8s resources, I want to use identical ones between all the environments. If I have 3 exposed services (e.g. web, api, magic) in each domain, routed by subdomain, I shouldn't require 3 separate ingresses, one for each environment, per exposed service. Instead, I want to do something like this:
>

```yml
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  name: web
spec:
  rules:
  - host: "web.*"
    http:
      paths:
      - path: /
        backend:
          serviceName: web-service
          servicePort: 80
```

**Comment by bowei** [comment]
We've had requests for the suffix wildcard. Unfortunately, this doesn't seem to interact well with CertManager that interprets the Host field as domains to register certificates.

**Comment by bowei** [comment]
It would be good to work through how those use case will work with the wildcard suffix. For example, DNS integrations look at the host field to register names. How will it work in that case, etc...

**Comment by deitch** [comment]
I am quite aware that `foo.*` is not compliant with the RFC, which is probably the certmanager issue. Lots of code depends on that.

The issue is that what we are trying to do here is very much with the cloud-native philosophy. If I need to maintain different `Ingress` configs for 3 different clusters, solely because they are in different environments (prod, qa, dev1, etc.) but are otherwise identical, that is a lot of overhead and boilerplate. They became more "pet-like".

**Comment by bowei** [comment]
That's a valid use case, but we need to be careful not to break existing integrations when we add functionality. For example, expanding the set of valid host paths requires a new API versions. There is the issue with being able to be supports by all implementations. It requires some thought, that's why it cannot be done unilaterally.

**Comment by gae123** [comment]
@rikatz would this allow for *-foo.bar.com match both a-foo.bar.com and b-foo.bar.com

This is what we need in one of the projects I am involved and this is why I have been following this ticket

**Comment by rikatz** [comment]
> @rikatz would the related 1.18 enhancement allow for *-foo.bar.com match both a-foo.bar.com and b-foo.bar.com
> 
> This is what we need in one of the projects I am involved and this is why I have been following this ticket

By my tests no. It supports *.domain.com or *.system.domain.com, but not a sub-registry wildcard (as *-a.example.com, *xpto.example.com).

**Comment by dannymk** [comment]
How about just "domain.com" as a catch all for 'www|myhost|myhost2|..."?  We definitely need this feature.

**Comment by crfrase** [comment]
> 
> 
> @deitch until then if anyone needs to do thins in the official nginx ingress, here is a shortcut:
> 
>     * Set `host: <just the subdomain>` in Ingress rule, e.g. `host: myapp`
> 
>     * And then use an init container in the nginx deployment (or daemonset) to replace `server_name {{$server.Hostname}};` in the default nginx template with `server_name {{$server.Hostname}} {{$server.Hostname}}.*;`

I was able to setup this patch in my controller deployment. It worked well for most cases, but the --enable-ssl-passthrough cases weren't responding the the shortened hostname ingress. I don't suppose anyone knows what else would need to be patched to make 'nginx.ingress.kubernetes.io/ssl-passthrough: "true"' cases react to an ingress using a hostname without a domain?

**Comment by volnei** [comment]
Any update here?

**Comment by kobenauf** [comment]
I was interested in this too, because I have many domains (and new ones added often) that will use the same site essentially, but they each one their own domain.  But @bowei poses a critical problem.  If certs were given out like this, web.somehacker.com could point to your site and get a cert and open you up for new types of security bugs.

**Comment by chrisjohnson** [comment]
I would not expect that certs need to follow suit

**Comment by cmluciano** [comment]
@deitch Please take a look at the recent wildcard support that we added in 1.18 & 1.19 and let us know if this helps your use-case.

https://kubernetes.io/blog/2020/04/02/improvements-to-the-ingress-api-in-kubernetes-1.18/#support-for-hostname-wildcards
https://opensource.googleblog.com/2020/09/kubernetes-ingress-goes-ga.html

**Comment by chrisjohnson** [comment]
Checking the description it seems to imply that it's not a full wildcard, only a suffix match

```Precise matches require that the http host header matches the Host setting. Wildcard matches require the http host header is equal to the suffix of the wildcard rule.```

For my use case, where it's the domain itself that I want to be wildcarded (i.e. `service-name.*` will match `service-name.dev.mysite.com` or `service-name.production.mysite.com`)

**Comment by silverkorn** [comment]
We might need to rename the issue title because everyone here, me included, is requesting at least a support for wildcards as suffix, e.g.:
- `www.*`
- dev-www.*
- `dev.www.*`

**Comment by chrisjohnson** [comment]
I think the title and original issue body still imply what we are looking for, specifically full wildcard support (with the example given showing that it's not just suffix matching). I think @cmluciano was just pointing to a new feature that may help with some of the needs of some people here, not necessarily implying that the issue was now solved.

**Comment by cmluciano** [comment]
I'm checking with others in the Ingress community to see if this is something that might be possible

**Comment by distributethe6ix** [comment]
@cmluciano Were you able to find out if this was possible? It looks like both @johnreytanquinco & @deitch are looking for a wildcard capability on the "top-level-domain". I don't see this anywhere in the Ingress documentation for Kubernetes. 
The docs say we can set a wildcard for the subdomain not not domain or TLD.

https://kubernetes.io/docs/concepts/services-networking/ingress/#hostname-wildcards

Example:
website.* could mean website.io, or website.org, website.com,

**Comment by pyramation** [comment]
running into this limitation as well. When I manually use Traefik I can provide `HostRegExp={subdomain:[a-z0-9-]+}.mydomain.com` https://github.com/traefik/traefik/issues/4219

**Comment by stevengibbszions** [comment]
Still no native support in the K8s Ingress Spec for regex/wildcards?  Is that accurate?

**Comment by thockin** [comment]
@bowei Can you give this a final update and then close if appropriate

**Comment by bowei** [comment]
This is where the PathType `PathTypeImplementationSpecific` comes in. Creating a portable re syntax between implementations is quite a fraught topic: there are many subtle difference in behavior between re engines (PCRE, Re2, ECMAScript etc).

For Ingress spec -- use of PathTypeImplementationSpecific allows for implementations to use their own syntax to specify that the expression is a regex. For example, I think nginx uses the `~` to denote that the path is a regex.

 TLDR: you should see what your specific implementation uses for dealing with regex.

Note that in the [Gateway API](https://gateway-api.sigs.k8s.io/), there is support for specifying that the path is a regex, but the specific behavior is implementation-specific.

/close

Please reopen if there are any additional questions.

**Comment by chrisjohnson** [comment]
If somebody gets the time to test this out and confirm that we can use suffix patterns such as `Host: foo.*` (using nginx ingress controller's `~` feature) would you please post back here with your sample config so we can learn from it and also to confirm that this ticket truly is ready to be closed?

**Comment by bowei** [comment]
For host, a suffix wildcard is not supported wrt to the definition:

see https://kubernetes.io/blog/2020/04/02/improvements-to-the-ingress-api-in-kubernetes-1.18/#support-for-hostname-wildcards

**Comment by chrisjohnson** [comment]
Then I'm confused why this ticket is being closed since it exists to request support for exactly that

**Comment by deitch** [comment]
I second that. I would want to reopen it (I opened the original issue), but I apparently have insufficient rights. Would a maintainer please reopen it? Much appreciated, thank you.

**Comment by bowei** [comment]
I would create a new issue for your specific use case -- the original title is "Allow RegExp or wildcard for ingress domain and/or path";

it is quite broad (note: the "and/or") and some of the functionality is possible now:

* wildcards
* regex via ImplementationSpecific

**Comment by chrisjohnson** [comment]
We already have several people gathered here, watching this ticket, who all seem to be in alignment on what we are after. I would rather not throw that away and start from scratch simply because the title may not be exactly perfect. Surely we can rename the title of tickets? The example requested right in the original ticket body is asking for exactly this thing, and the followup discussion is all centered around this behavior.

Further, I actually think the title is apt; requesting regexp or wildcard is still the desired behavior.

It's not true that kubernetes actually currently supports true wildcards in host; if it did, I could specify `host: foo.*` (which is a form of wildcard). It currently supports some subset of wildcards, the way I would describe the current implementation is supporting prefixes for the host field.

**Comment by bowei** [comment]
/reopen

We should fix the title of this issue to match the proposal, if the use case is more narrow.
Support to allow regexp in general in a portable way is not currently planned for the reason explained above. Implementations can do things on their own, but not in a way that will be guaranteed by the API.

The issue with suffix wildcards is that is does not interact well with some uses (certs issuance). Also, some providers cannot support suffix wildcarding and we don't (currently) have a notion of [conformance level](https://github.com/kubernetes-sigs/gateway-api/issues/530) like in the Gateway API to allow this to be specified.

**Comment by gp187** [comment]
What would be very useful is to match the subdomain to the service name. That would be super helpful. 

Someone visits app.example.com then the Ingress would route to service.name = app

```
apiVersion: networking.k8s.io/v1
kind: Ingress
  name: serviceX-i
  namespace: examples
spec:
  rules:
  - host: *.example.com # if dns  = app.example.com
    http:
      paths:
      - backend:
          service:
            name: $1 # then name is `app`
            port:
              number: 3010
        path: /
```

**Comment by thockin** [comment]
Ingress has `ImplementationSpecific` path-types, which is the only reasonable way to do this.

Ingress also supports wildcards for hostname matching.

Beyond that, Ingress as an API is effectively "done".  Further work needs to aim at the Gateway API https://gateway-api.sigs.k8s.io/

**Comment by chrisjohnson** [comment]
I think there's some misunderstanding, this feature has not been implemented. A path match does nothing to help with matching the host, `ImplementationSpecific` does not impact the `host: "web.*"` field.

Also, as mentioned 2 years ago](https://github.com/kubernetes/kubernetes/issues/41881#issuecomment-704406910) the hostname field does **not** support wildcards, it only supports **suffixes**. Here is an example wildcard that it does not support:

```
host: "service-name.*"
```

**Comment by thockin** [comment]
Sorry, I used the wrong code for closing this.

Further work needs to aim at the Gateway API https://gateway-api.sigs.k8s.io/

There's not much interest in adding non-portable extensions to Ingress when Gateway is just about here.

**Comment by ztec** [comment]
> There's not much interest in adding non-portable extensions to Ingress when Gateway is just about here.

I don't know yet Gateway API, will look into it, but I find it a bit harsh to cut any development and evolution to an established and used API just because a new one is about here.

Is Ingress API deprecated ?

**Comment by thockin** [comment]
Anything we add to Ingress API still needs to be implemented by controller impls, which have indicated very low desire to double-track such work.

If a significant corpus of Ingress implementations could easily support this, I'm not going to hard block it.  I would need to see that laid out, along with someone thinking thru the design here (probably not hard, similar to how `PathType` works I would guess (without thinking toooo hard on it).
